Add continuous integration (GitHub Actions)

marvinschopf opened the pull request:
Hello! 👋

I have developed a CI integration for `js.org`. (see [`js-org-validate`](https://github.com/marvinschopf/js-org-validate)).

It checks the following things:
- `cnames_active.js` file exists
- File contains valid JavaScript
- All entries are sorted alphabetically
- All entries are valid URLs or CNAMEs
- All URLs are reachable and return a valid HTTP status code (in this case, however, no error is generated, only a warning)

The GitHub Actions workflow does the following:
- Sets up a Node.js environment
- Runs the validator using `yarn dlx`.
- Generates a comment in the pull request or commit showing the details of the validation.

In addition, the validator also generates some nice statistics about the services used by the different pages (e.g.: GitHub Pages, Netlify, Vercel etc).

I've opened a few example PRs in my repository that show the behaviour of the validator:
- [Wrong sorting](https://github.com/marvinschopf/js.org/pull/11)
- [Missing file 'cnames_active.js'](https://github.com/marvinschopf/js.org/pull/10)
- [Invalid CNAME target url](https://github.com/marvinschopf/js.org/pull/9)
- [Multiple commits](https://github.com/marvinschopf/js.org/pull/8)
- [Invalid CNAME](https://github.com/marvinschopf/js.org/pull/7)
- [Invalid JS syntax](https://github.com/marvinschopf/js.org/pull/6)
- [Success](https://github.com/marvinschopf/js.org/pull/5)

Best regards
Marvin

MattIPv4 commented:
🎉 Awesome, this looks amazing! tyvm!

My one immediate request before I dive into this further would be that we bring the script into this repo as part of this PR, instead of relying on a published package, so that we can more easily maintain it as js.org going forward.

MattIPv4 commented:
I'd also be curious to know why we're using `pull_request_target` here and not `pull_request`? It seems like some extra logic is having to be baked in to handle `pull_request_target`, which I've not seen needed with `pull_request` historically using Actions.

marvinschopf commented:
Hey @MattIPv4!

Thanks for the quick answer!

`pull_request_target` is necessary because otherwise the workflow (or GITHUB_TOKEN) won't get the necessary `write` permissions to create a comment under the pull request.

However, moving the source code of the validator to this repository would create an attack vector, as a user could create a pull request with modified `test.js`. This script could then create, for example, a different `pull_request_comment.md` than intended, which would then be created unfiltered in the repository. Therefore, it is an additional security measure to keep the validator separate from the repository, as the repository is then safe from infiltration of malicious code by a pull request.

I would suggest creating a repository under the `js-org` organisation to store the validator. The workflow could then load the validator from this repository via GitHub Package Registry. Then maintainers could easily make changes to the validator and the code would still be safe from an attack.

MattIPv4 commented:
Ah, TIL, the perms that token gets are so jank it drives me insane -- `pull_request_target` makes sense then :)

Agree on the attack vector of running the script direct from the branch -- instead of creating a separate repo though (extra overhead, would be cleaner to have the validator code directly here without needing to worry about releases to npm or the package repo etc.), if the code is committed as part of this PR, the workflow can then specifically check out the master branch of this repo to get and execute the script?

marvinschopf commented:
All right, I have moved the validator to the repository and modified the workflow accordingly. It now copies the fork to a subfolder where the fork cannot do any sort of damage against the test script. In my eyes, it should run safely like this.

MattIPv4 commented:
Awesome, tysm -- I'll dive into the code tomorrow and get you a full review of it all asap :)

An initial thing that I imagine I'll bring up in said review will be that this currently seems to rely on one rather massive index file that's rather ungainly to mentally parse, it would be great if this could be broken up into individual concerns a bit more to make it easier to work with.

marvinschopf commented:
Yes, you're right there. I'll get to work tomorrow on making the whole thing a bit cleaner.

MattIPv4 commented on test/src/providers.ts:
```diff
@@ -0,0 +1,104 @@
+		return true;
+	if (url.match(/^(?:https?:\/\/)?(?:[^.]+\.)?gitbook\.com(\/.*)?$/gi))
+		return true;
+	return false;
+}
+
+function isGitlab(url: string): boolean {
+	if (url.match(/^(?:https?:\/\/)?(?:[^.]+\.)?gitlab\.io(\/.*)?$/gi))
+		return true;
+	return false;
+}
+
+function isReplit(url: string): boolean {
+	if (url.match(/^(?:https?:\/\/)?(?:[^.]+\.)?repl\.co(\/.*)?$/gi))
+		return true;
+	return false;
+}
+
+export function identify(url: string): string {
+	const _url: string = urlWithoutProtocol(url);
+	if (isVercel(_url)) {
+		return "Vercel";
+	}
+	if (isNetlify(_url)) {
+		return "Netlify";
+	}
+	if (isGithub(_url)) {
+		return "GitHub Pages";
+	}
+	if (isGitbook(_url)) {
+		return "Gitbook.com";
+	}
+	if (isGitlab(_url)) {
+		return "GitLab";
+	}
+	if (isReplit(_url)) {
+		return "Replit";
+	}
+	return "Other";
+}
```
A cname target is invalid if it has `https?://` at the start of it, so it seems like all the regex here to deal with that is complexity that we don't really need, and will make maintenance harder?

MattIPv4 reviewed (changes requested):
Tyvm once again 💙 -- some initial thoughts on the code:

marvinschopf commented on test/src/components/Status.tsx:
```diff
@@ -0,0 +1,50 @@
+				<Text color="cyan" bold>
+					{this.props.loading && (
+						<React.Fragment>
+							<Spinner type="dots" />{" "}
```
The spinner is only displayed when the script is executed in the terminal on a computer. It's just a nice UI addition:)

MattIPv4 commented:
👋 Hey @marvinschopf, just wanted to check in on how things were going with this and if there's anything we can do to help get this CI landed and merged?

marvinschopf commented:
Hi @MattIPv4, I've been very busy the last few days but will get back to it this week. 👍🏼

MattIPv4 commented:
Going to close this out as we merged validation using the existing js.org-cleanup logic in #7088 -- I encourage you to contribute to that repo though if you're interested in improving the validation beyond what it already handles :)
